File: LeetCode/28.find-the-index-of-the-first-occurrence-in-a-string.py
#
# @lc app=leetcode id=28 lang=python3
#
# [28] Find the Index of the First Occurrence in a String
#
# https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/description/
#
# algorithms
# Easy (44.66%)
# Likes:    6628
# Dislikes: 489
# Total Accepted:    3.2M
# Total Submissions: 7.2M
# Testcase Example:  '"sadbutsad"\n"sad"'
#
# Given two strings needle and haystack, return the index of the first
# occurrence of needle in haystack, or -1 if needle is not part of haystack.
# 
# 
# Example 1:
# 
# 
# Input: haystack = "sadbutsad", needle = "sad"
# Output: 0
# Explanation: "sad" occurs at index 0 and 6.
# The first occurrence is at index 0, so we return 0.
# 
# 
# Example 2:
# 
# 
# Input: haystack = "leetcode", needle = "leeto"
# Output: -1
# Explanation: "leeto" did not occur in "leetcode", so we return -1.
# 
# 
# 
# Constraints:
# 
# 
# 1 <= haystack.length, needle.length <= 10^4
# haystack and needle consist of only lowercase English characters.
# 
# 
#

# @lc code=start
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        # 获取 haystack 和 needle 的长度
        n, m = len(haystack), len(needle)

        # 如果 needle 为空，返回 0
        if m == 0:
            return 0

        # 遍历 haystack，检查每个子串是否等于 needle
        for i in range(n - m + 1):
            if haystack[i:i + m] == needle:
                return i

        # 如果没有找到，返回 -1
        return -1
# @lc code=end

File: 00LeetCodeCpp/28.find-the-index-of-the-first-occurrence-in-a-string.cpp
// ...existing code...
/*
 * @lc app=leetcode id=28 lang=cpp
 *
 * [28] Find the Index of the First Occurrence in a String
 */

// @lc code=start
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.empty()) return 0;
        int n = haystack.size(), m = needle.size();
        // 构建 LPS（最长相同前后缀）数组
        vector<int> lps(m, 0);
        for (int i = 1, len = 0; i < m; ++i) {
            while (len > 0 && needle[i] != needle[len]) len = lps[len - 1];
            if (needle[i] == needle[len]) ++len;
            lps[i] = len;
        }
        // KMP 搜索
        int j = 0; // needle 指针
        for (int i = 0; i < n; ++i) {
            while (j > 0 && haystack[i] != needle[j]) j = lps[j - 1];
            if (haystack[i] == needle[j]) ++j;
            if (j == m) return i - m + 1;
        }
        return -1;
    }
};
// @lc code=end
// ...existing code...